src/main/java/org/opensearch/neuralsearch/search/HitsThresholdChecker.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.neuralsearch.search;
+
+import java.util.Locale;
+
+import lombok.Getter;
+
+import org.apache.lucene.search.ScoreMode;
+
+/**
+ *  Abstracts algorithm that allows early termination for the search flow if number of hits reached
+ *  certain treshold
+ */
+public class HitsThresholdChecker {
+    private int hitCount;
+    @Getter
+    private final int totalHitsThreshold;
+
+    private HitsThresholdChecker(int totalHitsThreshold) {
+        if (totalHitsThreshold < 0) {
+            throw new IllegalArgumentException(String.format(Locale.ROOT, "totalHitsThreshold must be >= 0, got %d", totalHitsThreshold));
+        }
+        if (totalHitsThreshold == Integer.MAX_VALUE) {
+            throw new IllegalArgumentException(String.format(Locale.ROOT, "totalHitsThreshold must be less than max integer value"));
```

Comment:
What is wrong with `totalHitsThreshold` having `Integer.MAX_VALUE` but okay with `Integer.MAX_VALUE - 1`?

Reply:
we don't want to spill over and reset to negative, this what will happen on next increment if current value is ```Integer.MAX_VALUE```. ```Integer.MAX_VALUE - 1``` doesn't have such issue. 
For clarification, we're counting number of hits for one sub-query in one shard. Integer type should handle max number of hits (used by Lucene in similar class https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/HitsThresholdChecker.java)